src/bin/tikv-fail.rs

```diff
@@ -0,0 +1,163 @@
+                        .takes_value(true)
+                        .help(
+                            "Recover fail points. Eg. tikv-fail recover fail::a fail::b",
+                        ),
+                )
+                .arg(
+                    Arg::with_name("file")
+                        .short("f")
+                        .takes_value(true)
+                        .help("Recover from a file of fail points"),
+                ),
+        )
+        .subcommand(SubCommand::with_name("list").about("List all fail points"));
+    let matches = app.clone().get_matches();
+    let addr = matches.value_of("addr").unwrap();
+    let addr = addr.trim_left_matches("http://");
+
+    let env = Arc::new(EnvBuilder::new().name_prefix("tikv-fail").build());
+    let channel = ChannelBuilder::new(env).connect(addr);
+    let client = DebugClient::new(channel);
+
+    if let Some(matches) = matches.subcommand_matches("inject") {
+        let mut list = matches.value_of("file").map_or_else(Vec::new, read_file);
+        for pair in matches.values_of("args").unwrap() {
+            let mut parts = pair.split('=');
+            list.push((
+                parts.next().unwrap().to_owned(),
+                parts.next().unwrap_or("").to_owned(),
+            ))
+        }
+
+        for (name, actions) in list {
+            if actions.is_empty() {
+                println!("No action for fail point {}", name);
+                continue;
+            }
+            let mut inject_req = debugpb::InjectFailPointRequest::new();
+            inject_req.set_name(name);
+            inject_req.set_actions(actions);
+
+            let option = CallOption::default().timeout(Duration::from_secs(10));
+            client.inject_fail_point_opt(inject_req, option).unwrap();
+        }
+    } else if let Some(matches) = matches.subcommand_matches("recover") {
+        let mut list = matches.value_of("file").map_or_else(Vec::new, read_file);
+        for f in matches.values_of("args").unwrap() {
+            list.push((
+                f.to_owned(),
+                "".to_owned(),
+            ))
+        }
+
+        for (name, _) in list {
+            let mut recover_req = debugpb::RecoverFailPointRequest::new();
+            recover_req.set_name(name);
+
+            let option = CallOption::default().timeout(Duration::from_secs(10));
+            client.recover_fail_point_opt(recover_req, option).unwrap();
+        }
+    } else if matches.is_present("list") {
+        let list_req = debugpb::ListFailPointsRequest::new();
+        let option = CallOption::default().timeout(Duration::from_secs(10));
+        let resp = client.list_fail_points_opt(list_req, option).unwrap();
+        println!("{:?}", resp.get_entries());
+    }
+}
+
+fn read_file(path: &str) -> Vec<(String, String)> {
+    let mut list = vec![];
+    let mut buffer = String::new();
+    fs::File::open(path)
+        .unwrap()
+        .read_to_string(&mut buffer)
```

Comment:
Can you use `BufReader` instead? So you can collect lines without allocate memory twice the size. Read line lazily is better.